### PR TITLE
Default config edit scope

### LINKED ISSUE #2184: What config scope should be the default?
It came up while documenting #2152 that Spack is inconsistent about which config scope is edited from command to command.  Currently, the sort-of-inconsistently-implemented default behavior is to edit the highest-precedence available scope, which used to be `user`, but as of #2030 it's the platform-specific scope for the current platform (e.g. `~/.spack/bgq/config.yaml`).

I figure I should bounce this off of people before editing it.  See if you think what I'm proposing below is reasonable.

Currently there are 11 commands that take a scope argument.  I've labeled them as either "read" operations or "modify" operations.  I think they should behave as follows:

**Read operations:**
- `spack config get` 
- `spack compilers / spack compiler list`
- `spack mirror list`
- `spack repo list`

I think all of these should show merged results by default, and allow the user to select a specific scope with `--scope=<scope>` if they want to.

**Modify operations:**
- `spack config edit`
- `spack compiler add / spack compiler find`
- `spack compiler remove`
- `spack mirror add`
- `spack mirror remove`
- `spack repo add`
- `spack repo remove`

I think instead of taking the highest precedence scope (which among other things might change from spack version to spack version if we implement additional scopes), we should implement this policy:
1. Introduce a `default_edit_scope` option in `config.yaml`, and out of the box it would be set to edit the `user` scope by default.
2. If the option is set to `<scope>`, commands should edit the generic `<scope>` config by default, but it should edit an architecture-specific `<scope>/<platform>` config _if it exists_.  I think this is a) natural and b) prevents the user from getting confused if they, say, forgot they made a platform-specific config.
3. Finally, compilers are a special case (in that they are actually platform-specific) and `spack compiler add/find` should modify the platform-specific scope by default.

Does this policy seem reasonable?  the other option would be to require a scope parameter for edit commands, but I think it's nice for users to be able to say, e.g., `spack config edit` or `spack repo add` without thinking about scopes.

@alalazo @davydden @adamjstewart @citibeth

### PR DESCRIPTION
**Pull request by scheibelp**
Closes #2184

* the default edit scope is "user" (`~/.spack/`) unless there is an existing configuration file in the `user/$platform` scope (e.g. `~/.spack/darwin/compilers.yaml`)
* (EDIT 8/23 - forgot to mention) for the `compilers` section, the default edit scope is always `user/$platform`, even if no configuration file exists at the `user/$platform` scope
* add a `config.yaml` variable called `default_edit_scope` which lets users set where this occurs
* add a `$platform` config variable which expands to the platform from which Spack is being run: this allows users to explicitly set `default_edit_scope: user/$platform`
* Note that if a user sets any scope without including `$platform`, Spack will always try to use the corresponding platform scope if a file exists for it
* (EDIT: added) Remove the `spack.util.path` module and move all functions in it into `spack.config`
* (EDIT: added) updated tests to check these rules and the `$platform` variable expansion

TODOs:

- [x] Handle item (3) of #2184
- [ ] (?) should it be possible for users to indicate that they want to edit the non-platform version of a section?

### CODE REVIEW

File: lib/spack/spack/config.py [hunk 1]
```diff
@@ -557,12 +560,17 @@ def override(path_or_scope, value=None):
 command_line_scopes = []
 
 
+def _platform_scope(scope_name):
+    platform = spack.architecture.platform().name
+    return '%s/%s' % (scope_name, platform)
```
Comment (tldahlgren) [inline lib/spack/spack/config.py]: Should this be `os.path.join(scope_name, platform)`?

File: lib/spack/spack/config.py [hunk 2]
```diff
@@ -808,13 +816,35 @@ def they_are(t):
...
 #
-def default_modify_scope():
+def default_modify_scope(section):
     """Return the config scope that commands should modify by default.
 
     Commands that modify configuration by default modify the *highest*
     priority scope.
     """
-    return spack.config.config.highest_precedence_scope().name
+    default_edit_scope = spack.config.get('config:default_edit_scope')
```
Comment (tldahlgren) [inline lib/spack/spack/config.py]: FWIW. You could simplify lines 825-827 using:  `default_edit_scope = spack.config.get('config:default_edit_scope', 'user')`.

**Comment by tldahlgren** [comment]
Not sure what my assignment is on this.

LGTM from a coding perspective.  I'm kind of on the fence about moving `cannonicalize_path` from `spack.util.path` to `spack.config`, but I can appreciate that it does rely on configuration variable substitutions.

**Comment by adrienbernede** [comment]
Hello Peter, 
Thank you for tackling this "more complex than it may seem" problem.

> add a config.yaml variable called default_edit_scope which lets users set where this occurs

I have a concrete example of a situation I would like to be able to handle with scopes in Spack:
When using Spack in GitLab on a cluster, the runner will act under the UID of the user triggering the pipeline. However, I would like Spack to be "user-config" agnostic in that case, so that I am sure the behavior will be the same for any user.

Does this feature address that kind of situation ? and if yes, how ?

Thank you

**Comment by scheibelp** [comment]
Hi Adrien,

> When using Spack in GitLab on a cluster, the runner will act under the UID of the user triggering the pipeline. However, I would like Spack to be "user-config" agnostic in that case, so that I am sure the behavior will be the same for any user.
> Does this feature address that kind of situation ? and if yes, how ?

I don't think this feature would address your issue, but I do think Spack can already do what you desire: I think you would want to create a custom config directory, and when you run Spack you can refer to that directory with `-C` like `spack -C <the custom config directory>`; in order to ignore any user configuration, you would need to create a file for each section and use the `::` syntax described in https://spack.readthedocs.io/en/latest/configuration.html#overriding-entire-sections to ensure that all other scopes are ignored (this works because custom configuration scopes have the highest precedence as covered in https://spack.readthedocs.io/en/latest/configuration.html#platform-specific-scopes).

That should work, but it is onerous enough that I think it would merit a separate `--no-user-config` option - this PR does not provide that though.

**Comment by adrienbernede** [comment]
> That should work, but it is onerous

I agree with this statement, but not only because of the :: syntax to add in every configuration section, but also because I would have to specify the config path for every action, with no mistake allowed.

I would love to see "no-higher-config" as a key in config-file erasing previously parsed config though (overkill, unnatural).
Or site config being of higher precedence but user-config being the one modified by default (twisted).

Anyway, I’ve got the answer to my question.
Thanks ! 

PS:
https://spack.readthedocs.io/en/latest/configuration.html#scope-precedence

> Completely ignoring _higher-level_ (lower-level, right ?) configuration options is supported with the :: notation

**Comment by scheibelp** [comment]
> PS:
> https://spack.readthedocs.io/en/latest/configuration.html#scope-precedence
>> Completely ignoring higher-level (lower-level, right ?) configuration options is supported with the :: notation

I think it's referring to the ordering presented in https://spack.readthedocs.io/en/latest/configuration.html#platform-specific-scopes. The wording could certainly be improved to clarify that though!

**Comment by tldahlgren** [comment]
Is this still WIP?

**Comment by alalazo** [comment]
@scheibelp Looking at old PRs on my list. Can this be closed or is it still WIP?

**Comment by scheibelp** [comment]
I'll close it. It is so far out of sync that even though https://github.com/spack/spack/issues/2184 remains open (i.e. we want to avoid editing the `~` scope by default), I'd want to reimplement the idea from scratch. Other efforts may also mitigate this need (other work for example adds an install-tree scope and makes that the default, which would close this).

**Pull request closed without merging**